### DISCUSSION TITLE
4GB Upload

### DISCUSSION THREAD

**Discussion by silincone** [reo7sp/tgbot-cpp#303] [Q&A] [answered]
Is it possible to upload 4GB documents, if so, how?

**Comment by we-lu** [comment]
Information on sending files are written here: https://core.telegram.org/bots/api#sending-files

Either you
- upload the file manually to Telegram. Then you need Telegram Premium to upload files up to 4GB. Without Premium the limit is at 2GB
- upload the file via API to Telegram. Then you have the file limit somewhere at 20MB

**Reply by silincone** [reply]
I see, but look at [this](https://github.com/SilentDemonSD/WZML-X) repo for example, it is written in python and uses pyrogram framework to download/upload files to telegram.

Without premium account, the bot is able to download upto 4GB size document and upload 2GB size document, 
With premium account "string", it can upload & download upto 4GB size document. 

If Pyrogram can do it, why not TgBot?! I am just trying to find a c++ alternative to pyrogram, thought that TgBot would do, but it seems to not help. 

Please do reply back, I have no one else to ask this with.

I am a beginner at C++ and I do not know any other programming languages, so guess you'll understand me.

**Comment by reo7sp** [comment, marked as answer]
Pyrogram uses Telegram MTProto as a user client; tgbot-cpp wraps the HTTP Bot API, so the capabilities and limits are different.

For the hosted endpoint `https://api.telegram.org`:

- multipart upload: up to 50 MB for general files;
- download through `getFile`: up to 20 MB;
- sending an existing Telegram `file_id`: no re-upload limit.

A self-hosted `telegram-bot-api` server can download without a size limit and upload up to 2000 MB, not 4 GB. To use it with tgbot-cpp:

```cpp
TgBot::CurlHttpClient httpClient;
TgBot::Bot bot(token, httpClient, "http://localhost:8081");
```

That URL works only when a local Bot API server is actually running there. tgbot-cpp cannot bypass either server-side limit; 4 GB uploads through a Premium user account are an MTProto feature, not a Bot API feature.
